File: clients/haskell/src/Model/Game.hs
module Model.Game where

import Prelude hiding (id)
import qualified Trans
import Trans (Trans)
import Data.Int
import Model.Loot (Loot)
import Model.Player (Player)
import Model.Projectile (Projectile)
import Model.Sound (Sound)
import Model.Unit (Unit)
import Model.Zone (Zone)

-- | Current game's state
data Game = Game {
    -- | Your player's id
    myId :: Int32,
    -- | List of players (teams)
    players :: [Player],
    -- | Current tick
    currentTick :: Int32,
    -- | List of units visible by your team
    units :: [Unit],
    -- | List of loot visible by your team
    loot :: [Loot],
    -- | List of projectiles visible by your team
    projectiles :: [Projectile],
    -- | Current state of game zone
    zone :: Zone,
    -- | List of sounds heard by your team during last tick
    sounds :: [Sound] }
    deriving Show

instance Trans Game where
    read = do
        myId <- Trans.read
        players <- Trans.read
        currentTick <- Trans.read
        units <- Trans.read
        loot <- Trans.read
        projectiles <- Trans.read
        zone <- Trans.read
        sounds <- Trans.read
        return Game {
            myId,
            players,
            currentTick,
            units,
            loot,
            projectiles,
            zone,
            sounds }
    
    write Game {
        myId,
        players,
        currentTick,
        units,
        loot,
        projectiles,
        zone,
        sounds } = do
            Trans.write myId
            Trans.write players
            Trans.write currentTick
            Trans.write units
            Trans.write loot
            Trans.write projectiles
            Trans.write zone
            Trans.write sounds

File: clients/rust/src/model/game.rs
use super::*;

/// Current game's state
#[derive(Clone, Debug)]
pub struct Game {
    /// Your player's id
    pub my_id: i32,
    /// List of players (teams)
    pub players: Vec<model::Player>,
    /// Current tick
    pub current_tick: i32,
    /// List of units visible by your team
    pub units: Vec<model::Unit>,
    /// List of loot visible by your team
    pub loot: Vec<model::Loot>,
    /// List of projectiles visible by your team
    pub projectiles: Vec<model::Projectile>,
    /// Current state of game zone
    pub zone: model::Zone,
    /// List of sounds heard by your team during last tick
    pub sounds: Vec<model::Sound>,
}

impl trans::Trans for Game {
    fn write_to(&self, writer: &mut dyn std::io::Write) -> std::io::Result<()> {
        self.my_id.write_to(writer)?;
        self.players.write_to(writer)?;
        self.current_tick.write_to(writer)?;
        self.units.write_to(writer)?;
        self.loot.write_to(writer)?;
        self.projectiles.write_to(writer)?;
        self.zone.write_to(writer)?;
        self.sounds.write_to(writer)?;
        Ok(())
    }
    fn read_from(reader: &mut dyn std::io::Read) -> std::io::Result<Self> {
        let my_id: i32 = trans::Trans::read_from(reader)?;
        let players: Vec<model::Player> = trans::Trans::read_from(reader)?;
        let current_tick: i32 = trans::Trans::read_from(reader)?;
        let units: Vec<model::Unit> = trans::Trans::read_from(reader)?;
        let loot: Vec<model::Loot> = trans::Trans::read_from(reader)?;
        let projectiles: Vec<model::Projectile> = trans::Trans::read_from(reader)?;
        let zone: model::Zone = trans::Trans::read_from(reader)?;
        let sounds: Vec<model::Sound> = trans::Trans::read_from(reader)?;
        Ok(Self {
            my_id,
            players,
            current_tick,
            units,
            loot,
            projectiles,
            zone,
            sounds,
        })
    }
}